File: sort_examples/shellSort.js
/* eslint-disable @typescript-eslint/explicit-function-return-type */
// eslint-disable-next-line no-undef
sort = (array, film) => {
  let compares = 0
  for (
    let k = 2, step = Math.floor(array.length / 2);
    step > 0;
    k *= 2, step = Math.floor(array.length / k)
  ) {
    for (let i = step; i < array.length; i++) {
      const temp = array[i]
      let j
      for (j = i; j >= step; j -= step) {
        film.rec({ array, i, j, temp, compares })
        compares++
        if (array[j - step] > temp) {
          array[j] = array[j - step]
          film.rec({ array, i, j, temp, compares })
        } else {
          break
        }
      }
      array[j] = temp
      film.rec({ array, i, j, temp, compares })
    }
  }
}

File: sort_examples/shell_sort.py
def sort(array):
    gap = len(array) // 2
    while gap > 0:
        array.note("gap", gap)
        for i in range(gap, len(array)):
            temp = array[i]
            j = i
            while j >= gap and array[j - gap] > temp:
                array[j] = array[j - gap]
                j -= gap
            array[j] = temp
        gap //= 2

    array.mark_sorted(*range(len(array)))
